### PR TITLE
test: MBT: Add slashing to model

### PR DESCRIPTION
**Pull request by p-offtermatt**
## Description

Closes: #1531

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Provided a link to the relevant issue or specification
- [x] Reviewed "Files changed" and left comments if necessary 
- [x] Confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] Confirmed all author checklist items have been addressed
- [ ] Confirmed that this PR does not change production code

### CODE REVIEW

File: tests/mbt/model/ccv.qnt [hunk 1]
```diff
@@ -654,6 +685,141 @@ module ccv {
...
+        // TODO: implement jailing
+        // val newProviderState = currentState.providerState
+        //  .jailUntil(currentState.providerState, packet.validator, jailEndTime)
+        //  .slash(currentState.providerState, packet.validator, packet.valPower, slashFactorToBeDetermined)
+
+        // Ok(currentState.with("providerState", newProviderState))
+        Err("not implemented")
+    }
+
+    pure def recvVscMaturedPacketOnProvider(currentState: ProtocolState, sender: Chain, packet: VscMaturedPacket): Result = {
```
Comment (p-offtermatt) [inline tests/mbt/model/ccv.qnt:756]: moved from ccv_utils because the "sibling function" recvSlashPacketOnProvider needs to access a parameter from the module

File: tests/mbt/driver/generate_traces.sh [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
+go run ./... -modelPath=../model/ccv_boundeddrift.qnt -step stepBoundedDrift -invariant CanReceiveMaturations -traceFolder traces/bound_mat -numTraces 2 -numSteps 30 -numSamples 20
 echo "Generating synced traces with maturations"
-go run ./... -modelPath=../model/ccv_sync.qnt -init initSync -step stepSync -invariant CanReceiveMaturations -traceFolder traces/sync_mat -numTraces 1 -numSteps 300 -numSamples 20
+go run ./... -modelPath=../model/ccv_sync.qnt -init initSync -step stepSync -invariant CanReceiveMaturations -traceFolder traces/sync_mat -numTraces 2 -numSteps 30 -numSamples 20
 echo "Generating long synced traces without invariants"
-go run ./... -modelPath=../model/ccv_sync.qnt -init initSync -step stepSync -traceFolder traces/sync_noinv -numTraces 1 -numSteps 500 -numSamples 1
-go run ./... -modelPath=../model/ccv_boundeddrift.qnt --step stepBoundedDriftKeyAssignment --traceFolder traces/bound_key -numTraces 1 -numSteps 100 -numSamples 20
\ No newline at end of file
+go run ./... -modelPath=../model/ccv_sync.qnt -init initSync -step stepSync -traceFolder traces/sync_noinv -numTraces 2 -numSteps 30 -numSamples 1
```
Comment (p-offtermatt) [inline tests/mbt/driver/generate_traces.sh]: decreased numSteps here, since the traces were too large to open/work with. In turn, I increased the numTraces to 2, so it should still give some coverage but be easier to debug

File: tests/mbt/driver/generate_more_traces.sh [hunk 3]
```diff
@@ -3,11 +3,11 @@
 echo "Generating bounded drift traces with timeouts"
 go run ./... -modelPath=../model/ccv_boundeddrift.qnt -step stepBoundedDrift  -invariant CanTimeoutConsumer -traceFolder traces/bound_timeout -numTraces 20 -numSteps 200 -numSamples 200
 echo "Generating long bounded drift traces without invariants"
-go run ./... -modelPath=../model/ccv_boundeddrift.qnt -step stepBoundedDrift -traceFolder traces/bound_noinv -numTraces 20 -numSteps 500 -numSamples 1
+go run ./... -modelPath=../model/ccv_boundeddrift.qnt -step stepBoundedDrift -traceFolder traces/bound_noinv -numTraces 20 -numSteps 300 -numSamples 1
```
Comment (p-offtermatt) [inline tests/mbt/driver/generate_more_traces.sh:6]: Similar reasoning, traces were just too large to work with

**Comment by p-offtermatt** [comment]
Note: I will add some more info to the README (specifically, that the slash meter is not modelled aka assumed to be infinite)

File: tests/mbt/driver/setup.go [hunk 4]
```diff
@@ -107,7 +108,7 @@ func getAppBytesAndSenders(
 
 	// Create genesis accounts.
 	for i := 0; i < len(nodes); i++ {
-		pk := secp256k1.GenPrivKey()
+		pk := secp256k1.GenPrivKeyFromSecret([]byte(chainID + valNames[i]))
```
Comment (p-offtermatt) [inline tests/mbt/driver/setup.go:111]: Made to not rely on os randomness to get reproducible runs

File: testutil/integration/validators.go [hunk 5]
```diff
@@ -17,7 +26,7 @@ func CreateValidators(n int) (
 		signersByAddress = make(map[string]tmtypes.PrivValidator, n)
 	)
 	for i := 0; i < n; i++ {
-		privVal := mock.NewPV()
+		privVal := mock.PV{PrivKey: ed25519.GenPrivKeyFromSecret([]byte(chainId + fmt.Sprint(i)))}
```
Comment (p-offtermatt) [inline testutil/integration/validators.go:29]: Changed to make this deterministic. This caused some other unrelated files to need to change.

File: tests/mbt/model/ccv.qnt [hunk 6]
```diff
@@ -607,11 +793,11 @@ module ccv {
                 }
 
                 // check whether the validator has positive power
-                pure val provValSet = currentState.providerState.chainState.currentValidatorSet
+                pure val provValSet = currentState.providerState.chainState.currentValidatorPowers
                 pure val provValPower = if (provValSet.keys().contains(providerNode)) provValSet.get(providerNode) else 0
                 pure val consumersWithAddrAssignmentChangesInThisBlock = 
-                    if (provValPower > 0) {
-                        // if the consumer has positive power, the relevant key assignment for the consumer changed
+                    if (provValPower > 0 and providerNode.in(nonJailedNodes(currentState.providerState))) {
```
Comment (sainoe) [inline tests/mbt/model/ccv.qnt:799]: Why don't we set the voting power of jailed validators to `0`?

File: tests/mbt/model/ccv.qnt [hunk 7]
```diff
@@ -654,6 +840,168 @@ module ccv {
...
+    }
+
+    pure def recvSlashPacketOnProvider(currentState: ProtocolState, sender: Chain, packet: SlashPacket): Result = {
+        // determine the jail duration for the validator:
+        // if the packet is a downtime slash, the jail duration is DowntimeJailTime
+        // otherwise, the validator is jailed until the DoubleSignJailTime
+        val jailEndTime = if (packet.downtime) {
+            packet.sendingTime + DowntimeJailTime
+        } else {
+            packet.sendingTime + DoubleSignJailEndTime
```
Comment (sainoe) [inline tests/mbt/model/ccv.qnt:893]: The provider is only handling `SlashPackets` for downtime infractions. Packets with double-signing infractions are dropped on receipt.
See https://github.com/cosmos/interchain-security/blob/main/x/ccv/provider/keeper/relay.go#L313-L329.

File: tests/mbt/model/ccv.qnt [hunk 8]
```diff
@@ -654,6 +840,168 @@ module ccv {
...
+            // if there is no entry for that validator, there is no key assignment
+            .getOrElse(packet.validator, packet.validator)
+
+        val slashFactor = if (packet.downtime) {
+            DowntimeSlashPercentage
+        } else {
+            DoubleSignSlashPercentage
+        }
+        
+        val newProviderState =
```
Comment (sainoe) [inline tests/mbt/model/ccv.qnt:908]: Shouldn't we check that the validator isn't already jailed before?

File: tests/mbt/model/ccv.qnt (hunk 6 above)
Comment (p-offtermatt) [reply tests/mbt/model/ccv.qnt:799]: Because this makes unjailing operations hard, which we might want to add later. I filter the jailed validators out when the valset is entered on-chain during EndBlock

File: tests/mbt/model/ccv.qnt (hunk 8 above)
Comment (p-offtermatt) [reply tests/mbt/model/ccv.qnt:908]: Great point - I didn't catch this because I don't compare the jail times, will add this check in the driver and also adjust this here

File: tests/mbt/model/ccv.qnt (hunk 7 above)
Comment (sainoe) [reply tests/mbt/model/ccv.qnt]: Oh yea, I missed it. Yes please, it isn't clear that `DoubleSignJailEndTime` and `DoubleSignSlashPercentage` are never used to update the provider states.

File: tests/mbt/model/ccv.qnt [hunk 9]
```diff
@@ -662,6 +848,161 @@ module ccv {
...
+        currentState: ProtocolState,
+        consumer: Chain,
+        consumerNode: Node,
+        vscId: VscId,
+        downtime: bool): Result = {
+        val consumerState = currentState.consumerStates.get(consumer)
+        if (not(isRunningConsumer(consumer, currentState.providerState))) {
+            Err("Consumer is not currently a consumer - must have 'running' status!")
+        } else if (downtime and consumerState.outstandingDowntime.contains(consumerNode)) {
+            Err("Consumer already has outstanding downtime for validator!")
```
Comment (sainoe) [inline tests/mbt/model/ccv.qnt:865]: In ICS, this condition is just perform an no-op. Are we simulating the same behaviour by returning an error here?

Comment (p-offtermatt) [reply tests/mbt/model/ccv.qnt:865]: An error means the model will not permit doing the action, and no trace will ever exhibit this behaviour.
I'll change that to be an actual noop - thanks for the diligence, great catch!

File: tests/mbt/model/ccv_model.qnt [hunk 10]
```diff
@@ -948,4 +1060,163 @@ module ccv_model {
                 VotingPowerChange("node1", 50)
             }
         )
+
+    // ======= tests for consumer initiated slashing =======
+
+    run DoubleDowntimeSlashTest =
```
Comment (sainoe) [inline tests/mbt/model/ccv_model.qnt:1066]: Nit: For what "DoubleDowntime" stands for?

File: tests/mbt/model/ccv_model.qnt [hunk 11]
```diff
@@ -438,6 +466,13 @@ module ccv_model {
             }
         )
 
+    // for all consumers, waitingForSlachPacketAck is either empty or has a single element
```
Comment (sainoe) [inline tests/mbt/model/ccv_model.qnt:469]:
```suggestion
    // for all consumers, waitingForSlashPacketAck is either empty or has a single element
```

File: tests/mbt/model/ccv.qnt [hunk 12]
```diff
@@ -662,6 +848,162 @@ module ccv {
...
+            // for the sender chain of this slash packet and indicate that we need to send a packet at endblock
+            val newDowntimeSlashRequests = newProviderState.downtimeSlashRequests
+                .getOrElse(sender, List())
+                .append(packet.validator)
+            val newProviderState2 = newProviderState.with(
+                "downtimeSlashRequests", newProviderState.downtimeSlashRequests.put(sender, newDowntimeSlashRequests)
+            ).with(
+                "consumersWithPowerChangesInThisEpoch", 
+                if (valSetChanged) {
+                    // valSet changes for all consumers, not only the sende,r because the validator will be jailed for everyone
```
Comment (sainoe) [inline tests/mbt/model/ccv.qnt:931]:
```suggestion
                    // valSet changes for all consumers, not only the sender because the validator will be jailed for everyone
```

File: tests/mbt/model/ccv_model.qnt (hunk 10 above)
Comment (p-offtermatt) [reply tests/mbt/model/ccv_model.qnt:1066]: Good call, at some point this tested two downtime slashes in the same trace, but was changed. I'll rename it

**Review by sainoe — APPROVED** [review]
Great work!

File: tests/mbt/model/ccv.qnt [hunk 13]
```diff
@@ -662,6 +848,162 @@ module ccv {
...
+
+            val slashFactor = DowntimeSlashPercentage
+            
+            val newProviderState = currentState.providerState
+                .jailUntil(providerVal, jailEndTime)
+                // // slashing is currently not enabled in ICS 
+                // .slash(packet.validator, packet.valPower, slashFactor)
+
+            // if the validator did not have voting power of 0 or was already jailed before, the validator set will change due to the slash
+            val valSetChanged = packet.downtime and
```
Comment (mpoke) [inline tests/mbt/model/ccv.qnt:918]: `packet.downtime` will always be true as it was checked above in the if condition.

**Review by mpoke — APPROVED** [review]
LGTM. Great work @p-offtermatt

File: tests/mbt/model/ccv.qnt [hunk 14]
```diff
@@ -81,7 +113,10 @@ module ccv_types {
             // Stores, for each consumer chain, the list of packets that have been sent to the consumer chain
             // and *have not been received yet*.
             // In the implementation, this would roughly be the unacknowledged packets on an ibc channel.
-            outstandingPacketsToConsumer: Chain -> List[VscPacket],
+            outstandingPacketsToConsumer: Chain -> List[Packet],
+
+            // Stores the packets that the provider will send an ack for when it ends the current block.
+            acksToSendOnEndBlock: Chain -> List[Packet],
```
[the remark this answers is not in the document]
Comment (p-offtermatt) [reply tests/mbt/model/ccv.qnt:119]: The code might send the acks immediately, but recall that they go through IBC, which will not process the acks before the block is ended.
